src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -244,7 +252,10 @@ private RangeIterator rangeMatch(Expression expression, AbstractBounds<Partition
             upperInclusive = false;
         }
 
-        Collector cd = new Collector(keyRange);
+        var capacity = Math.max(MINIMUM_QUEUE_SIZE, lastQueueSize.get());
+        var mergingIteratorBuilder = MergingRangeIterator.builder(keyBounds, indexContext.keyFactory(), capacity);
+        lastQueueSize.set(mergingIteratorBuilder.size());
```

Comment:
This is somewhat outside the scope of your PR, but I want to mention it in case it might be helpful. I wonder how accurate this `lastQueueSize` is. It seems like it could be pretty inaccurate, especially if queries have capture vastly different trie entries. It seems like it'd be valuable to have a way to estimate the number of leaves between two bounds of the trie. That logic along with an estimate on number of keys per leaf seems like it could be used for an a priori query planner that doesn't need to build these iterators in the first place.

Reply:
I thought the same. Honestly I don't like this "heuristic"; I even removed it for some time. It doesn't seem to bring a lot of performance win either, I noticed it increased the throughput by at most 5%, not sure even if it's not a noise. Anyway, I didn't want to change too much at once.

Reply:
> Anyway, I didn't want to change too much at once.

Works for me.